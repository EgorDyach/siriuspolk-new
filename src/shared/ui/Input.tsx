import { cx } from 'class-variance-authority';
import * as React from 'react';
import { Label } from './Label';

interface InputProps extends React.ComponentProps<'input'> {
  label?: React.ReactNode;
  error?: string;
}

function Input({
  className,
  required,
  label,
  type,
  name,
  error,
  ...props
}: InputProps) {
  return (
    <>
      <Label
        data-testid="label"
        htmlFor={name}
        className={cx(
          'mb-2 text-[12px] font-normal flex flex-col items-start xl:text-base',
          className,
        )}
      >
        <div className="text-left w-full flex justify-between items-end">
          <span>
            {required && (
              <span data-testid="required-mark" className="text-red-400">
                *
              </span>
            )}
            {label}
          </span>
          {error && (
            <span className="ml-2 text-end text-red-400 text-[12px] xl:text-[14px]">
              {error}
            </span>
          )}
        </div>
        <input
          name={name}
          type={type}
          data-slot="input"
          data-testid="input"
          required={required}
          className={cx(
            'file:text-foreground placeholder:text-[#B3B3B3] selection:bg-primary selection:text-primary-foreground dark:bg-input/30 border-black border-[1px] flex text-[10px] w-full min-w-32  bg-transparent text-base shadow-xs transition-[color,box-shadow] outline-none file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium disabled:pointer-events-none disabled:border-[#999] disabled:cursor-not-allowed disabled:opacity-50 p-[5px]',
            'focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[2px]',
            'aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive',
            '2sm:p-3 2sm:text-base',
            'xl:p-4 2sm:text-3xl xl:border-2',
          )}
          {...props}
        />
      </Label>
    </>
  );
}

export { Input };
