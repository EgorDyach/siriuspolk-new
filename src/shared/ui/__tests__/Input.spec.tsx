import { cleanup, render, screen } from '@testing-library/react';
import { Input } from '../Input';

describe('<Input />', () => {
  it('should render correctly', () => {
    render(<Input />);
    expect(screen.getByTestId('input')).toBeInTheDocument();
  });
  it('should render correctly with classname', () => {
    render(<Input className="test-class" />);
    expect(screen.getByTestId('label')).toHaveClass('test-class');
  });
  it('should render correctly with required', () => {
    render(<Input required />);
    expect(screen.getByTestId('input')).toBeRequired();
    expect(screen.getByTestId('required-mark')).toBeInTheDocument();
  });
  it('should render correctly with different types', () => {
    render(<Input type="email" />);
    expect(screen.getByTestId('input')).toHaveAttribute('type', 'email');
    cleanup();
    render(<Input type="password" />);
    expect(screen.getByTestId('input')).toHaveAttribute('type', 'password');
    cleanup();
    render(<Input type="file" />);
    expect(screen.getByTestId('input')).toHaveAttribute('type', 'file');
  });
  it('should render correctly with label', () => {
    render(<Input label="Test input" />);
    expect(screen.getByTestId('input')).toBeInTheDocument();
    expect(screen.getByText(/test input/i)).toBeInTheDocument();
  });
  it('should render correctly with error', () => {
    render(<Input error="Test error" />);
    expect(screen.getByTestId('input')).toBeInTheDocument();
    expect(screen.getByText(/test error/i)).toBeInTheDocument();
  });
  it('should connect name and htmlFor', () => {
    render(<Input name="test-name" />);
    expect(screen.getByTestId('input')).toHaveAttribute('name', 'test-name');
    expect(screen.getByTestId('label')).toHaveAttribute('for', 'test-name');
  });
  it('should render correctly with other props', () => {
    render(<Input disabled style={{ color: 'red' }} placeholder="test" />);
    expect(screen.getByTestId('input')).toBeDisabled();
    expect(screen.getByTestId('input')).toHaveAttribute('placeholder', 'test');
    expect(screen.getByTestId('input')).toHaveStyle('color: red');
  });
  it('matches snapshot with default input', () => {
    const { asFragment } = render(<Input />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('matches snapshot with label and required', () => {
    const { asFragment } = render(<Input label="Имя" required />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('matches snapshot with error', () => {
    const { asFragment } = render(<Input error="Ошибка" />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('matches snapshot with disabled', () => {
    const { asFragment } = render(<Input disabled />);
    expect(asFragment()).toMatchSnapshot();
  });
});
