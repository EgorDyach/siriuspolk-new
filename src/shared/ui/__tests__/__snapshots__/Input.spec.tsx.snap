// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> matches snapshot with default input 1`] = `
<DocumentFragment>
  <label
    class="flex items-center gap-1 text-sm leading-none font-medium select-none group-data-[disabled=true]:pointer-events-none group-data-[disabled=true]:opacity-50 peer-disabled:cursor-not-allowed peer-disabled:opacity-50 mb-2 text-[12px] font-normal flex flex-col items-start xl:text-base"
    data-slot="label"
    data-testid="label"
  >
    <div
      class="text-left w-full flex justify-between items-end"
    >
      <span />
    </div>
    <input
      class="file:text-foreground placeholder:text-[#B3B3B3] selection:bg-primary selection:text-primary-foreground dark:bg-input/30 border-black border-[1px] flex text-[10px] w-full min-w-32  bg-transparent text-base shadow-xs transition-[color,box-shadow] outline-none file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium disabled:pointer-events-none disabled:border-[#999] disabled:cursor-not-allowed disabled:opacity-50 p-[5px] focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[2px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive 2sm:p-3 2sm:text-base xl:p-4 2sm:text-3xl xl:border-2"
      data-slot="input"
      data-testid="input"
    />
  </label>
</DocumentFragment>
`;

exports[`<Input /> matches snapshot with disabled 1`] = `
<DocumentFragment>
  <label
    class="flex items-center gap-1 text-sm leading-none font-medium select-none group-data-[disabled=true]:pointer-events-none group-data-[disabled=true]:opacity-50 peer-disabled:cursor-not-allowed peer-disabled:opacity-50 mb-2 text-[12px] font-normal flex flex-col items-start xl:text-base"
    data-slot="label"
    data-testid="label"
  >
    <div
      class="text-left w-full flex justify-between items-end"
    >
      <span />
    </div>
    <input
      class="file:text-foreground placeholder:text-[#B3B3B3] selection:bg-primary selection:text-primary-foreground dark:bg-input/30 border-black border-[1px] flex text-[10px] w-full min-w-32  bg-transparent text-base shadow-xs transition-[color,box-shadow] outline-none file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium disabled:pointer-events-none disabled:border-[#999] disabled:cursor-not-allowed disabled:opacity-50 p-[5px] focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[2px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive 2sm:p-3 2sm:text-base xl:p-4 2sm:text-3xl xl:border-2"
      data-slot="input"
      data-testid="input"
      disabled=""
    />
  </label>
</DocumentFragment>
`;

exports[`<Input /> matches snapshot with error 1`] = `
<DocumentFragment>
  <label
    class="flex items-center gap-1 text-sm leading-none font-medium select-none group-data-[disabled=true]:pointer-events-none group-data-[disabled=true]:opacity-50 peer-disabled:cursor-not-allowed peer-disabled:opacity-50 mb-2 text-[12px] font-normal flex flex-col items-start xl:text-base"
    data-slot="label"
    data-testid="label"
  >
    <div
      class="text-left w-full flex justify-between items-end"
    >
      <span />
      <span
        class="ml-2 text-end text-red-400 text-[12px] xl:text-[14px]"
      >
        Ошибка
      </span>
    </div>
    <input
      class="file:text-foreground placeholder:text-[#B3B3B3] selection:bg-primary selection:text-primary-foreground dark:bg-input/30 border-black border-[1px] flex text-[10px] w-full min-w-32  bg-transparent text-base shadow-xs transition-[color,box-shadow] outline-none file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium disabled:pointer-events-none disabled:border-[#999] disabled:cursor-not-allowed disabled:opacity-50 p-[5px] focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[2px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive 2sm:p-3 2sm:text-base xl:p-4 2sm:text-3xl xl:border-2"
      data-slot="input"
      data-testid="input"
    />
  </label>
</DocumentFragment>
`;

exports[`<Input /> matches snapshot with label and required 1`] = `
<DocumentFragment>
  <label
    class="flex items-center gap-1 text-sm leading-none font-medium select-none group-data-[disabled=true]:pointer-events-none group-data-[disabled=true]:opacity-50 peer-disabled:cursor-not-allowed peer-disabled:opacity-50 mb-2 text-[12px] font-normal flex flex-col items-start xl:text-base"
    data-slot="label"
    data-testid="label"
  >
    <div
      class="text-left w-full flex justify-between items-end"
    >
      <span>
        <span
          class="text-red-400"
          data-testid="required-mark"
        >
          *
        </span>
        Имя
      </span>
    </div>
    <input
      class="file:text-foreground placeholder:text-[#B3B3B3] selection:bg-primary selection:text-primary-foreground dark:bg-input/30 border-black border-[1px] flex text-[10px] w-full min-w-32  bg-transparent text-base shadow-xs transition-[color,box-shadow] outline-none file:inline-flex file:h-7 file:border-0 file:bg-transparent file:text-sm file:font-medium disabled:pointer-events-none disabled:border-[#999] disabled:cursor-not-allowed disabled:opacity-50 p-[5px] focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[2px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive 2sm:p-3 2sm:text-base xl:p-4 2sm:text-3xl xl:border-2"
      data-slot="input"
      data-testid="input"
      required=""
    />
  </label>
</DocumentFragment>
`;
